src/psyclone/psyad/adjoint_visitor.py

```diff
@@ -290,10 +290,13 @@ def loop_node(self, node):
                                   (node.stop_expr, "upper bound"),
                                   (node.step_expr, "step")]:
             for ref in expr.walk(Reference):
+                if (isinstance(ref.parent, Call) and
```

Comment:
My thinking here is that almost all `for ref in expr.walk(Reference):` are wrong because they miss references, e.g. kind in literal ` 0.1_my_ref` is missed, or this new condition that I added will miss "i" in `call my_aos(i)%my_func(a,b,c)` (which is no worst than before the PR). I think almost all of them should be replaced with expr.reference_accesses(var_info, types) with types being if we have only READ/WRITE or also compile time symbols like call symbols, types, ...

But the reference_access needs a bit more work before this. I will create an issue but not add a TODO unless you disagree, because it needs to be updated in more places that I touched in this PR.

Reply:
It's issue #2542

Reply:
That sounds sensible Sergi. Please could you add a TODO though to the places where I've complained about duplication?

Reply:
Done